### PR TITLE
Tracking doc: rewrite `@wdio/static-server-service` package into TypeScript

### PR DESCRIPTION
**Pull request by suniljaiswal01**
This is a tracking issue for organising the effort of re-writing the code base from vanilla JS into TypeScript. If you are reading this and are interested getting involved, this is a great chance to do so. Here are the rules:

if you are interested converting source files from /packages/wdio-static-server-service into TypeScript please comment on this thread and mention the files you want to convert to avoid duplicate work
there is no reason to convert the whole package at once - smaller PRs make it easier to review the code
if you make a PR, reference this issue so we can keep track of the progress
if you make a PR, label it with TypeScript and the package name wdio-static-server-service so we can easier find all PRs
if you make a PR, assign it to the TypeScript Rewrite project

#5847

### CODE REVIEW

**Comment by suniljaiswal01** [comment]
This is ready for review

File: package.json [hunk 1]
```diff
@@ -66,6 +66,9 @@
     "@types/jest": "^26.0.0",
     "@typescript-eslint/eslint-plugin": "^4.1.0",
     "@typescript-eslint/parser": "^4.1.0",
+    "@types/express": "^4.17.8",
```
Comment (mgrybyk) [inline package.json:69]: do we need to have it globally? Can we move types to the package itself?

File: packages/wdio-static-server-service/src/launcher.ts [hunk 2]
```diff
@@ -10,14 +10,27 @@ const log = logger('@wdio/static-server-service')
 
 const DEFAULT_LOG_NAME = 'wdio-static-server-service.log'
 
+export interface MiddleWareOption{
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/src/launcher.ts:13]: can you please re-format the file properly?

File: packages/wdio-static-server-service/src/launcher.ts [hunk 3]
```diff
@@ -10,14 +10,27 @@ const log = logger('@wdio/static-server-service')
...
+    mount:string;
+    middleware:  string;
+}
+
+export interface FolderOption{
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/src/launcher.ts:18]: I believe we can move types to separate file

**Review by mgrybyk — CHANGES_REQUESTED** [review]
Mind dealing with tests as well?

File: packages/wdio-static-server-service/src/launcher.ts (hunk 3 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts:18]: should i move it to "static-server-service.d.ts" file? or should i just create a new file only for types?

Comment (christian-bromann) [reply packages/wdio-static-server-service/src/launcher.ts:18]: I did latter in my PRs

File: package.json (hunk 1 above)
Comment (suniljaiswal01) [reply package.json]: Done

File: packages/wdio-static-server-service/src/launcher.ts (hunk 3 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts]: Changes Done

File: packages/wdio-static-server-service/src/launcher.ts (hunk 2 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts]: Done...Please check

File: packages/wdio-static-server-service/static-server-service.d.ts [hunk 4]
```diff
@@ -2,7 +2,7 @@ declare module WebdriverIO {
     interface ServiceOption extends StaticServerOptions {}
 }
 
-interface StaticServerOptions {
+export interface StaticServerOptions {
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/static-server-service.d.ts:5]: can you avoid adding export here?

File: packages/wdio-static-server-service/tests/launcher.test.ts [hunk 5]
```diff
@@ -6,6 +6,7 @@ import StaticServerLauncher from '../src/launcher'
 
 jest.mock('fs-extra')
 
+
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/tests/launcher.test.ts:9]: remove

File: packages/wdio-static-server-service/tests/launcher.test.ts [hunk 6]
```diff
@@ -14,7 +15,7 @@ test('should not start server when no mount folder is defined', async () => {
 
 test('should be able to start server', async () => {
     const service = new StaticServerLauncher({
-        folders: { mount: 'foo', path: 'bar' }
+        folders: { mount: 'foo', path: 'bar' },
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/tests/launcher.test.ts:18]: no need to add an extra comma

File: packages/wdio-static-server-service/package.json [hunk 7]
```diff
@@ -30,10 +30,13 @@
     "@wdio/logger": "6.4.7",
     "express": "^4.14.0",
     "fs-extra": "^9.0.0",
-    "morgan": "^1.7.0"
+    "morgan": "^1.7.0",
+    "@types/express": "^4.17.8",
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/package.json:34]: I think we can move types to dev dependencies

File: packages/wdio-static-server-service/package.json [hunk 8]
```diff
@@ -30,10 +30,13 @@
...
+    "@types/fs-extra": "^9.0.1",
+    "@types/morgan": "^1.9.1"
   },
   "publishConfig": {
     "access": "public"
   },
   "types": "static-server-service.d.ts"
-}
+}
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/package.json:42]: let's keep the existing style of EOL if possible

File: package.json [hunk 9]
```diff
@@ -103,4 +103,4 @@
     }
   },
   "dependencies": {}
-}
+}
```
Comment (mgrybyk) [inline package.json:106]: let's keep the existing style of EOL if possible

**Review by mgrybyk — CHANGES_REQUESTED** [review]
Can you please fix failing tests?

File: packages/wdio-static-server-service/static-server-service.d.ts (hunk 4 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/static-server-service.d.ts]: removed export

File: packages/wdio-static-server-service/tests/launcher.test.ts (hunk 6 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/tests/launcher.test.ts]: Changes Done

File: packages/wdio-static-server-service/tests/launcher.test.ts (hunk 5 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/tests/launcher.test.ts]: Removed

File: packages/wdio-static-server-service/package.json (hunk 7 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/package.json]: Done

Comment (suniljaiswal01) [reply packages/wdio-static-server-service/package.json]: Done

File: packages/wdio-static-server-service/static-server-service.d.ts (hunk 4 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/static-server-service.d.ts]: Done

File: packages/wdio-static-server-service/tests/launcher.test.ts (hunk 5 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/tests/launcher.test.ts]: Removed

File: packages/wdio-static-server-service/tests/launcher.test.ts (hunk 6 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/tests/launcher.test.ts]: done

**Comment by suniljaiswal01** [comment]
Let me know if any other changes are required

File: packages/wdio-static-server-service/src/@types/wdio-logger-stub.d.ts [hunk 10]
```diff
@@ -0,0 +1 @@
+declare module '@wdio/logger';
```
Comment (christian-bromann) [inline packages/wdio-static-server-service/src/@types/wdio-logger-stub.d.ts:1]: This is not needed anymore, please rebase on top of latest master and remove this file

File: packages/wdio-static-server-service/package.json (hunk 8 above)
Comment (christian-bromann) [reply packages/wdio-static-server-service/package.json:44]: +1 , if you install the `editorconfig` plugin for your IDE this should go away

**Review by christian-bromann — CHANGES_REQUESTED** [review]

File: packages/wdio-static-server-service/src/@types/wdio-logger-stub.d.ts (hunk 10 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/@types/wdio-logger-stub.d.ts]: Done

File: packages/wdio-static-server-service/package.json (hunk 8 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/package.json]: Done

**Review by christian-bromann — APPROVED** [review]
LGTM 👍 

@mgrybyk any outstanding comments?

File: packages/wdio-static-server-service/src/launcher.ts [hunk 11]
```diff
@@ -1,43 +1,48 @@
...
 const log = logger('@wdio/static-server-service')
 
 const DEFAULT_LOG_NAME = 'wdio-static-server-service.log'
 
 export default class StaticServerLauncher {
-    constructor ({ folders, port = 4567, middleware = [] }) {
+    folders: FolderOption[] | null
+    port: number
+    middleware: MiddleWareOption[]
+    server: any
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/src/launcher.ts:18]: Is it possible to replace any with something meaningful?

File: packages/wdio-static-server-service/src/launcher.ts [hunk 12]
```diff
@@ -1,43 +1,48 @@
...
 import logger from '@wdio/logger'
+import { FolderOption, MiddleWareOption } from './types'
 
 const log = logger('@wdio/static-server-service')
 
 const DEFAULT_LOG_NAME = 'wdio-static-server-service.log'
 
 export default class StaticServerLauncher {
-    constructor ({ folders, port = 4567, middleware = [] }) {
+    folders: FolderOption[] | null
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/src/launcher.ts:15]: Are you sure it can be null?

File: packages/wdio-static-server-service/src/launcher.ts [hunk 13]
```diff
@@ -1,43 +1,48 @@
...
-        this.folders.forEach((folder) => {
-            log.info('Mounting folder `%s` at `%s`', path.resolve(folder.path), folder.mount)
+        this.folders.forEach((folder: FolderOption) => {
+            log.info('Mounting folder `%s` at `%s`', resolve(folder.path), folder.mount)
             this.server.use(folder.mount, express.static(folder.path))
         })
 
         this.middleware.forEach(
-            (ware) => this.server.use(ware.mount, ware.middleware))
+            (ware: MiddleWareOption) => this.server.use(ware.mount, ware.middleware))
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/src/launcher.ts:46]: If we really need it can you replace any with a meaningful type?

File: packages/wdio-static-server-service/tests/launcher.test.ts [hunk 14]
```diff
@@ -64,5 +64,5 @@ test('should register middlewares', async () => {
 })
 
 afterEach(() => {
-    express.mockClear()
+    (express as unknown as jest.Mock).mockClear()
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/tests/launcher.test.ts:67]: Is it necessary to use `as` twice?

**Review by mgrybyk — CHANGES_REQUESTED** [review]

File: packages/wdio-static-server-service/tests/launcher.test.ts (hunk 14 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/tests/launcher.test.ts:67]: If not using "as unknown", i am getting the below error:

`Conversion of type 'typeof e' to type 'Mock<any, any>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first`

File: packages/wdio-static-server-service/src/launcher.ts (hunk 12 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts:15]: Yes, as per the below line:

https://github.com/suniljaiswal01/webdriverio/blob/ts-static-server-service/packages/wdio-static-server-service/src/launcher.ts#L20

File: packages/wdio-static-server-service/src/launcher.ts (hunk 11 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts:18]: this.server = express() > this is where the variable is getting initialized.  We can replace any with "express.Express" , will that is meaningful ? or should we keep it as any

Comment (mgrybyk) [reply packages/wdio-static-server-service/src/launcher.ts:18]: I suppose so

File: packages/wdio-static-server-service/src/launcher.ts (hunk 13 above)
Comment (mgrybyk) [reply packages/wdio-static-server-service/src/launcher.ts:46]: I think I missed line

File: packages/wdio-static-server-service/src/launcher.ts (hunk 11 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts]: Changes done

File: packages/wdio-static-server-service/src/launcher.ts [hunk 15]
```diff
@@ -44,7 +44,8 @@ export default class StaticServerLauncher {
         this.middleware.forEach(
             (ware: MiddleWareOption) => this.server.use(ware.mount, ware.middleware))
 
-        await promisify(this.server.listen.bind(this.server))(this.port)
+        const listenServer = this.server.listen.bind(this.server)
```
Comment (suniljaiswal01) [inline packages/wdio-static-server-service/src/launcher.ts:46]: is this implementation correct?

Comment (mgrybyk) [reply packages/wdio-static-server-service/src/launcher.ts:46]: Yes, but next line is missing await

File: packages/wdio-static-server-service/src/launcher.ts [hunk 16]
```diff
@@ -1,45 +1,50 @@
...
             this.server.use(folder.mount, express.static(folder.path))
         })
 
         this.middleware.forEach(
-            (ware) => this.server.use(ware.mount, ware.middleware))
+            (ware: MiddleWareOption) => this.server.use(ware.mount, ware.middleware as unknown as express.Application))
 
-        await promisify(this.server.listen.bind(this.server))(this.port)
+        const listenServer = this.server.listen.bind(this.server)
+        listenServer(this.port)
```
Comment (mgrybyk) [inline packages/wdio-static-server-service/src/launcher.ts:47]: Missing await

**Review by mgrybyk — CHANGES_REQUESTED** [review]

File: packages/wdio-static-server-service/src/launcher.ts (hunk 15 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts:46]: when using promisify, i am getting issue on next line.

image

Comment (mgrybyk) [reply packages/wdio-static-server-service/src/launcher.ts:46]: Well, it should be used with promisify, investigate how to deal with that nicely. I believe we aren't the first ones trying that

Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts:46]: @mgrybyk adding the below code in launcher.ts file resolves the issue

declare module 'util' {
    function promisify<T1>(fn: (arg1: T1, callback: (err?: any) => void) => void): (arg1: T1) => Promise<void>;
}

The same has been suggested here :

https://github.com/microsoft/TypeScript/issues/26048#issuecomment-409973174

is this ok? or should i search for something else?

**Comment by suniljaiswal01** [comment]
@mgrybyk  test are failing for "wdio-shared-store-service" package

image

File: packages/wdio-static-server-service/src/launcher.ts [hunk 17]
```diff
@@ -1,43 +1,52 @@
-import path from 'path'
-import { promisify } from 'util'
+import { join, resolve } from 'path'
+
+declare module 'util' {
+    function promisify<T1>(fn: (arg1: T1, callback: (err?: any) => void) => void): (arg1: T1) => Promise<void>;
+}
```
Comment (christian-bromann) [inline packages/wdio-static-server-service/src/launcher.ts:5]: Why is this needed?

Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts:5]: There seems to be some issue with promisify overloading , similar issue is mentioned here :
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20497

By using the above code, we are specifying which method to overload instead of the generic one

**Comment by christian-bromann** [comment]
@mgrybyk any outstanding issues here or is it good to go?

**Comment by mgrybyk** [comment]
Let's fix the failing tests

File: packages/wdio-static-server-service/src/launcher.ts (hunk 15 above)
Comment (mgrybyk) [reply packages/wdio-static-server-service/src/launcher.ts]: can you set type for listen server variable instead of declaring module util?

See https://github.com/webdriverio/webdriverio/compare/ts-selenium-standlone-service-fix-types#diff-97c9d8dd32f5509001124f081429f24eR44

File: packages/wdio-static-server-service/src/launcher.ts (hunk 17 above)
Comment (mgrybyk) [reply packages/wdio-static-server-service/src/launcher.ts:5]: let's remove declare module util

**Comment by mgrybyk** [comment]
> @mgrybyk test are failing for "wdio-shared-store-service" package
> 
> image

maybe you need to merge latest master? I'm not sure, the tests are not failing in other branches

**Comment by suniljaiswal01** [comment]
> > @mgrybyk test are failing for "wdio-shared-store-service" package
> > image
> 
> maybe you need to merge latest master? I'm not sure, the tests are not failing in other branches

It is getting pass on my local machine as well, not sure why it is failing in pipeline

**Comment by christian-bromann** [comment]
> It is getting pass on my local machine as well, not sure why it is failing in pipeline

Can you try to rebase on top of master and re-build the project with `npm run setup-full`?

**Comment by suniljaiswal01** [comment]
> > It is getting pass on my local machine as well, not sure why it is failing in pipeline
> 
> Can you try to rebase on top of master and re-build the project with `npm run setup-full`?

Now that test is passing here as well, failure is because of some other issue

**Comment by christian-bromann** [comment]
> failure is because of some other issue

Yes, seems to be related to the package you are rewriting

```
    packages/wdio-static-server-service/src/launcher.ts:46:15 - error TS2322: Type '(arg1: any, arg2: (() => void) | undefined) => Promise<any>' is not assignable to type '(port: number) => Promise<any>'.

    46         const listen: (port: number) => Promise<any> = promisify(this.server.listen)
                     ~~~~~~
```

File: packages/wdio-static-server-service/src/launcher.ts (hunk 15 above)
Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts]: Follow the above step, but getting the below error on runtime:

`packages/wdio-static-server-service/src/launcher.ts:46:15 - error TS2322: Type '(arg1: any, arg2: (() => void) | undefined) => Promise<any>' is not assignable to type '(port: number) => Promise<any>'.

    46         const listen: (port: number) => Promise<any> = promisify(this.server.listen)`

**Comment by christian-bromann** [comment]
@suniljaiswal01 anything I can do to help on this?

**Comment by suniljaiswal01** [comment]
> @suniljaiswal01 anything I can do to help on this?

I am not able to resolve the below issue:

```
    packages/wdio-static-server-service/src/launcher.ts:46:15 - error TS2322: Type '(arg1: any, arg2: (() => void) | undefined) => Promise<any>' is not assignable to type '(port: number) => Promise<any>'.

    46         const listen: (port: number) => Promise<any> = promisify(this.server.listen)
```

File: packages/wdio-static-server-service/src/launcher.ts [hunk 18]
```diff
@@ -1,45 +1,50 @@
...
+            log.info('Mounting folder `%s` at `%s`', resolve(folder.path), folder.mount)
             this.server.use(folder.mount, express.static(folder.path))
         })
 
         this.middleware.forEach(
-            (ware) => this.server.use(ware.mount, ware.middleware))
+            (ware: MiddleWareOption) => this.server.use(ware.mount, ware.middleware as unknown as express.Application))
 
-        await promisify(this.server.listen.bind(this.server))(this.port)
+        const listen: (port: number) => Promise<any> = promisify(this.server.listen)
```
Comment (christian-bromann) [inline packages/wdio-static-server-service/src/launcher.ts:46]: this should fix the issue.
```suggestion
        const listen = promisify(this.server.listen)
```

Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts:46]: If i do the above changes then i am getting the below error:

```
(property) StaticServerLauncher.port: number
Expected 0 arguments, but got 1
```

Comment (suniljaiswal01) [reply packages/wdio-static-server-service/src/launcher.ts:46]: using this ` const listen = <(port: number) => Promise<any>> promisify(this.server.listen)`. is this Ok?

**Review by christian-bromann — APPROVED** [review]
👍  nice, good job finding a fix for this!
